exonum/src/runtime/migrations.rs

```diff
@@ -12,22 +12,144 @@
+//! # Migrations Overview
+//!
+//! The goal of a data migration is to prepare data of an Exonum service for use with an updated
+//! version of the service business logic. In this sense, migrations fulfil the same role
+//! as migrations in traditional database management systems.
+//!
+//! Migrations are performed via [`MigrationScript`]s, which are essentially wrappers around
+//! a closure. A script takes data of a service and uses the [database capabilities] to transform
+//! it to a new version. Migration is non-destructive, i.e., does not remove the old versions
+//! of migrated indexes. Instead, new indexes are created in a separate namespace, and atomically
+//! replace the old data when the migration is flushed.
+//! (See [database docs][database capabilities] for more details.)
+//!
+//! The problems solved by the migration workflow are:
+//!
+//! - Allowing for migration to be performed in background, while the node continues to process
+//!   transactions and other requests.
+//! - Ensuring that migrations finish at finite time (i.e., at some blockchain height).
+//! - Allowing concurrent migrations for different services.
+//! - Ensuring that all nodes in the network have arrived at the same data after migration
+//!   is completed.
+//!
+//! Similar to other service lifecycle events, data migrations are managed by the [dispatcher],
+//! but are controlled by the [supervisor service].
+//!
+//! # Migration Workflow
+//!
+//! For a migration to start, the targeted service must be stopped, and a newer version of
+//! the service artifact needs to be deployed across the network and considered active.
+//!
+//! 1. Migration is *initiated* by a call from a supervisor. Once a block with this call is merged,
+//!   all nodes in the network retrieve the migration script via [`Runtime::migrate()`]
+//!   and start executing it in a background thread. The script may execute at varying speed
+//!   on different nodes. Service status changes to [`Migrating`].
+//!
+//! 2. After the script is finished on a node, its result becomes available using
+//!   the [`local_migration_result()`] method of the dispatcher schema. Nodes agree these results
+//!   using supervisor capabilities (e.g., via broadcasting transactions).
+//!
+//! 3. Once the consensus is built up around migration, its result is either *committed* or
+//!   the migration is *rolled back*. Right below, we consider commitment workflow; the rollback
+//!   workflow will be described slightly later.
+//!
+//! 4. Committing a migration works similarly to [artifact commitment]. It means that any node
+//!   in the network starting from a specific blockchain height must have migration completed
+//!   with a specific outcome (i.e., hash of the migrated data). A node that does not have
+//!   migration script completed by this moment will block until the script is completed.
+//!   If the local migration outcome differs from the committed one, the node will be unable
```

Comment:
The outcome of this scenario is described in the transaction rollback workflow below.

Reply:
Ah, I get it now. I think an additional note that lack of consensus causes rollback. Initially I understood the text in point 3 as "We should get the consensus, and the decision in consensus is either commit or rollback", while (as I get it now) it's rather "If we 1) get the consensus and 2) the result is successful for consensus participants, then we commit the migration; otherwise we do a rollback".